service/history/mutableStateBuilder.go

```diff
@@ -246,6 +240,67 @@ func (e *mutableStateBuilder) CloseUpdateSession() (*mutableStateSessionUpdates,
 	return updates, nil
 }
 
+func (e *mutableStateBuilder) assignEventIdToBufferedEvents() {
+	newCommittedEvents := e.hBuilder.history
+
+	for i, event := range newCommittedEvents {
+		if event.GetEventId() != bufferedEventID {
+			continue
+		}
+
+		eventID := e.executionInfo.NextEventID
+		event.EventId = common.Int64Ptr(eventID)
+		e.executionInfo.NextEventID++
+
+		switch event.GetEventType() {
+		case workflow.EventTypeActivityTaskStarted:
+			// need to update startedEventId in mutableState if it is still there
+			scheduledEventId := event.ActivityTaskStartedEventAttributes.GetScheduledEventId()
+			if ai, ok := e.GetActivityInfo(scheduledEventId); ok {
```

Comment:
Can you check if creating 2 updates in the same transaction for same activityInfo will work?